src/services/DealService.ts

```diff
@@ -108,7 +110,52 @@ export class DealService {
     /**
      * deals will take care of themselves on account changes
      */
-    this.getDeals();
+    if (!this.deals) {
+      await this.getDeals();
+    }
+
+    this.observeDeals();
+  }
+
+  private async observeDeals(): Promise<void> {
+    this.deals = new Map<Address, DealTokenSwap>();
```

Comment:
do you need to set `this.initializing = true` like getDeals does?  How do you handle reentrancy?